### PR TITLE
rhobot artifacts

### PR DESCRIPTION
**Pull request by frank0631**
new command line feature. fetching of artifacts from gocd

### CODE REVIEW

File: gocd/gocd.go [hunk 1]
```diff
@@ -286,6 +286,124 @@ func (server Server) pipelineGET(pipelineName string) (pipeline Pipeline, etag s
...
+		err = fmt.Errorf("Bad response code: %d with response: %s", resp.StatusCode, body)
+		return
+	}
+
+	err = json.Indent(&historyJSON, body, "", "\t")
+	if err != nil {
+		log.Warn("Failed to prettify JSON: ", err)
+	}
+
+	// log.Debug("pipelineHistory JSON:", string(prettyJSON.Bytes()))
```
Comment (tanderegg) [inline gocd/gocd.go]: Let's leave the debug in there

File: gocd/gocd.go [hunk 2]
```diff
@@ -286,6 +286,124 @@ func (server Server) pipelineGET(pipelineName string) (pipeline Pipeline, etag s
...
+	return
+}
+
+// History gets the run history of a pipeline of a given name exist. returns map
+func History(server *Server, name string) (latestRuns map[string]int, err error) {
```
Comment (tanderegg) [inline gocd/gocd.go:351]: Let's add some more comments to better annotate what's going on (for others and our future selves)

File: gocd/gocd_test.go [hunk 3]
```diff
@@ -86,6 +86,34 @@ func TestGocdGET(t *testing.T) {
 	}
 }
 
+// //Test require a pipeline to have a history
```
Comment (tanderegg) [inline gocd/gocd_test.go]: Add in comment here briefly explaining why these are commented and what we could do to test in the future.

File: rhobot.go [hunk 4]
```diff
@@ -66,6 +67,26 @@ func main() {
...
+	}
+	artifactPathFlag := cli.StringFlag{
+		Name:  "artifact-path",
+		Value: "cruise-output/console.log",
+		Usage: "which artifact to get",
+	}
+	artifactSavePathFlag := cli.StringFlag{
+		Name:  "save",
+		Value: "",
+		Usage: "which run of the stage history",
```
Comment (tanderegg) [inline rhobot.go]: Copy.paste error :)

File: rhobot.go [hunk 5]
```diff
@@ -126,6 +147,65 @@ func main() {
...
+				if c.Args().Get(0) != "" && c.Args().Get(1) != "" && c.Args().Get(2) != "" {
+					pipeline = c.Args().Get(0)
+					stage = c.Args().Get(1)
+					job = c.Args().Get(2)
+				} else {
+					log.Error("You must provide the PIPELINE STAGE and JOB of the artifact")
+					return
+				}
+
+				if c.String(pipelineRunFlag.GetName()) != "" {
```
Comment (tanderegg) [inline rhobot.go:184]: I think this approach is preferable to the two below

File: rhobot.go [hunk 6]
```diff
@@ -315,3 +396,59 @@ func healthcheckRunner(config *config.Config, healthcheckPath string, reportPath
...
+	artifactBuffer, err := gocd.Artifact(gocdServer, pipeline, pipelineRunNum, stage, stageRunNum, job, artifactPath)
+	if err != nil {
+		log.Fatalf("Failed to fetch artifact: %v", artifactPath)
+	}
+
+	//write to file or log
+	if artifactSavePath == "" {
+		artifactBuffer.WriteTo(os.Stdout)
+	} else {
+		f, createErr := os.Create(artifactSavePath)
```
Comment (tanderegg) [inline rhobot.go]: Sadly, we must catch each error sequentially.

**Review by tanderegg — CHANGES_REQUESTED** [review]
Great work Frank, just a few changes!

File: gocd/gocd.go (hunk 1 above)
Comment (frank0631) [reply gocd/gocd.go]: added it back in 21360bb9931065514aeb581cf0c7282830c65dae, along with a refractor to limit rewritten code

File: gocd/gocd.go (hunk 2 above)
Comment (frank0631) [reply gocd/gocd.go:351]: better comments in c7a93a5

File: gocd/gocd_test.go (hunk 3 above)
Comment (frank0631) [reply gocd/gocd_test.go]: better comments in f9acaa6

File: rhobot.go (hunk 6 above)
Comment (frank0631) [reply rhobot.go]: separated in e4fcb0d

File: rhobot.go (hunk 5 above)
Comment (frank0631) [reply rhobot.go:184]: changed in b0b38d9

File: rhobot.go (hunk 4 above)
Comment (frank0631) [reply rhobot.go]: changed in 7168cd3

**Review by tanderegg — APPROVED** [review]

**Pull request merged**